project-addons/custom_account/account_treasury_forecast.py

```diff
@@ -0,0 +1,157 @@
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from openerp import models, fields, api, exceptions, _
+
+
+PAYMENT_MODE = [('debit_receipt', 'Debit receipt'),
+                ('transfer', 'Transfer'),
+                ('both', 'Both')]
+
+
+class AccountTreasuryForecast(models.Model):
+    _inherit = "account.treasury.forecast"
+
+    payment_mode_customer = fields.Selection(PAYMENT_MODE, 'Payment mode', default='both')
+    account_bank = fields.Many2one('res.partner.bank', 'Account bank',
+                                   domain=lambda self: [('partner_id', '=', self.env.user.company_id.partner_id.id)])
+    check_old_open_customer = fields.Boolean(string="Old (opened)")
+    opened_start_date_customer = fields.Date(string="Start Date")
+    payment_mode_supplier = fields.Selection(PAYMENT_MODE, 'Payment mode', default='both')
+    check_old_open_supplier = fields.Boolean(string="Old (opened)")
+    opened_start_date_supplier = fields.Date(string="Start Date")
+
+    @api.one
+    @api.constrains('payment_mode_customer', 'check_old_open_customer',
+                    'payment_mode_supplier', 'check_old_open_supplier',
+                    'opened_start_date_customer', 'opened_start_date_supplier', 'start_date')
+    def check_filter(self):
+        if not self.payment_mode_customer or not self.payment_mode_supplier:
+            raise exceptions.Warning(
+                _('Error!:: You must select one option for payment mode fields.'))
+        elif self.payment_mode_customer != 'debit_receipt':
+            if self.check_old_open_customer:
+                if self.opened_start_date_customer >= self.start_date:
+                    raise exceptions.Warning(
+                        _('Error!:: Start date of old opened invoices in customers must be lower '
+                          'than the start date specified before.'))
+        elif self.payment_mode_supplier != 'debit_receipt':
+            if self.check_old_open_supplier:
+                if self.opened_start_date_supplier >= self.start_date:
+                    raise exceptions.Warning(
+                        _('Error!:: Start date of old opened invoices in suppliers must be lower '
+                          'than the start date specified before.'))
+
+    @api.one
+    def calculate_invoices(self):
+        invoice_obj = self.env['account.invoice']
+        treasury_invoice_obj = self.env['account.treasury.forecast.invoice']
+        new_invoice_ids = []
+        in_invoice_lst = []
+        out_invoice_lst = []
+
+        # CUSTOMER
+        search_filter_customer = ['&', ('type', 'in', ['out_invoice', 'out_refund'])]
```

Comment:
¿Estáis seguros de que este & no provocará fallos en ningún caso? Para 'both' lo veo bien, pero en otros casos es posible que no, por ejemplo con 'transfer', ¿probastéis todas las posibles casuísticas?

Reply:
Sí, lo he probado y no da problema porque, en cualquier caso, se añaden más condiciones.

Reply:
ok